// @wordpress/ui/build-module/tabs/context.mjs
// packages/ui/src/tabs/context.tsx
import {
  createContext,
  useContext,
  useCallback,
  useMemo,
  useRef,
  useEffect
} from "@wordpress/element";
import { Fragment, jsx } from "react/jsx-runtime";
var VALIDATION_ENABLED = process.env.NODE_ENV !== "production";
var TabsValidationContext = VALIDATION_ENABLED ? createContext(null) : null;
function useRegisterTabDev() {
  const context = useContext(TabsValidationContext);
  useEffect(() => {
    if (context) {
      return context.registerTab();
    }
    return void 0;
  }, [context]);
}
function useRegisterTabProd() {
}
var useRegisterTab = VALIDATION_ENABLED ? useRegisterTabDev : useRegisterTabProd;
function useRegisterPanelDev() {
  const context = useContext(TabsValidationContext);
  useEffect(() => {
    if (context) {
      return context.registerPanel();
    }
    return void 0;
  }, [context]);
}
function useRegisterPanelProd() {
}
var useRegisterPanel = VALIDATION_ENABLED ? useRegisterPanelDev : useRegisterPanelProd;
function TabsValidationProviderDev({
  children
}) {
  const tabCountRef = useRef(0);
  const panelCountRef = useRef(0);
  const validationScheduledRef = useRef(null);
  const scheduleValidation = useCallback(() => {
    if (validationScheduledRef.current) {
      clearTimeout(validationScheduledRef.current);
    }
    validationScheduledRef.current = setTimeout(() => {
      const tabCount = tabCountRef.current;
      const panelCount = panelCountRef.current;
      if (tabCount !== panelCount) {
        throw new Error(
          `Tabs: Tab/Panel count mismatch (${tabCount} Tabs, ${panelCount} Panels). Each Tab must be associated with exactly one Panel. Mismatched or missing associations can break screen reader navigation and violate WAI-ARIA Tabs pattern requirements.`
        );
      }
      validationScheduledRef.current = null;
    }, 0);
  }, []);
  const registerTab = useCallback(() => {
    tabCountRef.current += 1;
    scheduleValidation();
    return () => {
      tabCountRef.current -= 1;
      scheduleValidation();
    };
  }, [scheduleValidation]);
  const registerPanel = useCallback(() => {
    panelCountRef.current += 1;
    scheduleValidation();
    return () => {
      panelCountRef.current -= 1;
      scheduleValidation();
    };
  }, [scheduleValidation]);
  useEffect(() => {
    return () => {
      if (validationScheduledRef.current) {
        clearTimeout(validationScheduledRef.current);
      }
    };
  }, []);
  const contextValue = useMemo(
    () => ({
      registerTab,
      registerPanel
    }),
    [registerTab, registerPanel]
  );
  return /* @__PURE__ */ jsx(TabsValidationContext.Provider, { value: contextValue, children });
}
function TabsValidationProviderProd({
  children
}) {
  return /* @__PURE__ */ jsx(Fragment, { children });
}
var TabsValidationProvider = VALIDATION_ENABLED ? TabsValidationProviderDev : TabsValidationProviderProd;
export {
  TabsValidationProvider,
  useRegisterPanel,
  useRegisterTab
};
