// @wordpress/ui/build-module/alert-dialog/popup.mjs
// packages/ui/src/alert-dialog/popup.tsx
import { forwardRef, useContext } from "@wordpress/element";
import { __ } from "@wordpress/i18n";
import { Button } from "../button/index.mjs";
import * as Dialog from "../dialog/index.mjs";
import { AlertDialogContext } from "./context.mjs";

// packages/ui/src/alert-dialog/style.module.css
if (typeof document !== "undefined" && process.env.NODE_ENV !== "test" && !document.head.querySelector("style[data-wp-hash='35295c8521']")) {
  const style = document.createElement("style");
  style.setAttribute("data-wp-hash", "35295c8521");
  style.appendChild(document.createTextNode("@layer wp-ui-utilities, wp-ui-components, wp-ui-compositions, wp-ui-overrides;@layer wp-ui-compositions{._2ddc2fc9155a1dad__irreversible-action{--wp-ui-button-background-color:var(--wpds-color-bg-interactive-error-strong,#cc1818);--wp-ui-button-background-color-active:var(--wpds-color-bg-interactive-error-strong-active,#b90000);--wp-ui-button-foreground-color:var(--wpds-color-fg-interactive-error-strong,#f2efef);--wp-ui-button-foreground-color-active:var(--wpds-color-fg-interactive-error-strong-active,#f2efef)}}"));
  document.head.appendChild(style);
}
var style_default = { "irreversible-action": "_2ddc2fc9155a1dad__irreversible-action" };

// packages/ui/src/alert-dialog/popup.tsx
import { jsx, jsxs } from "react/jsx-runtime";
var Popup2 = forwardRef(
  function AlertDialogPopup({
    title,
    children,
    onConfirm,
    confirmButtonText = __("OK"),
    cancelButtonText = __("Cancel"),
    loading
  }, ref) {
    const { intent } = useContext(AlertDialogContext);
    const ConfirmButton = loading !== void 0 ? Button : Dialog.Action;
    return /* @__PURE__ */ jsxs(Dialog.Popup, { ref, children: [
      /* @__PURE__ */ jsx(Dialog.Header, { children: /* @__PURE__ */ jsx(Dialog.Title, { children: title }) }),
      children,
      /* @__PURE__ */ jsxs(Dialog.Footer, { children: [
        /* @__PURE__ */ jsx(
          Dialog.Action,
          {
            variant: "minimal",
            disabled: loading || void 0,
            children: cancelButtonText
          }
        ),
        /* @__PURE__ */ jsx(
          ConfirmButton,
          {
            className: intent === "irreversible" ? style_default["irreversible-action"] : void 0,
            onClick: onConfirm,
            loading,
            children: confirmButtonText
          }
        )
      ] })
    ] });
  }
);
export {
  Popup2 as Popup
};
