// @wordpress/ui/build/empty-state/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/ui/src/empty-state/index.ts
var empty_state_exports = {};
__export(empty_state_exports, {
  Actions: () => import_actions.Actions,
  Description: () => import_description.Description,
  Icon: () => import_icon.Icon,
  Root: () => import_root.Root,
  Title: () => import_title.Title,
  Visual: () => import_visual.Visual
});
module.exports = __toCommonJS(empty_state_exports);
var import_root = require("./root.cjs");
var import_visual = require("./visual.cjs");
var import_icon = require("./icon.cjs");
var import_title = require("./title.cjs");
var import_description = require("./description.cjs");
var import_actions = require("./actions.cjs");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Actions,
  Description,
  Icon,
  Root,
  Title,
  Visual
});
