// @wordpress/ui/build/alert-dialog/popup.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/ui/src/alert-dialog/popup.tsx
var popup_exports = {};
__export(popup_exports, {
  Popup: () => Popup2
});
module.exports = __toCommonJS(popup_exports);
var import_element = require("@wordpress/element");
var import_i18n = require("@wordpress/i18n");
var import_button = require("../button/index.cjs");
var Dialog = __toESM(require("../dialog/index.cjs"));
var import_context = require("./context.cjs");

// packages/ui/src/alert-dialog/style.module.css
if (typeof document !== "undefined" && process.env.NODE_ENV !== "test" && !document.head.querySelector("style[data-wp-hash='35295c8521']")) {
  const style = document.createElement("style");
  style.setAttribute("data-wp-hash", "35295c8521");
  style.appendChild(document.createTextNode("@layer wp-ui-utilities, wp-ui-components, wp-ui-compositions, wp-ui-overrides;@layer wp-ui-compositions{._2ddc2fc9155a1dad__irreversible-action{--wp-ui-button-background-color:var(--wpds-color-bg-interactive-error-strong,#cc1818);--wp-ui-button-background-color-active:var(--wpds-color-bg-interactive-error-strong-active,#b90000);--wp-ui-button-foreground-color:var(--wpds-color-fg-interactive-error-strong,#f2efef);--wp-ui-button-foreground-color-active:var(--wpds-color-fg-interactive-error-strong-active,#f2efef)}}"));
  document.head.appendChild(style);
}
var style_default = { "irreversible-action": "_2ddc2fc9155a1dad__irreversible-action" };

// packages/ui/src/alert-dialog/popup.tsx
var import_jsx_runtime = require("react/jsx-runtime");
var Popup2 = (0, import_element.forwardRef)(
  function AlertDialogPopup({
    title,
    children,
    onConfirm,
    confirmButtonText = (0, import_i18n.__)("OK"),
    cancelButtonText = (0, import_i18n.__)("Cancel"),
    loading
  }, ref) {
    const { intent } = (0, import_element.useContext)(import_context.AlertDialogContext);
    const ConfirmButton = loading !== void 0 ? import_button.Button : Dialog.Action;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(Dialog.Popup, { ref, children: [
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Dialog.Header, { children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Dialog.Title, { children: title }) }),
      children,
      /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(Dialog.Footer, { children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          Dialog.Action,
          {
            variant: "minimal",
            disabled: loading || void 0,
            children: cancelButtonText
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          ConfirmButton,
          {
            className: intent === "irreversible" ? style_default["irreversible-action"] : void 0,
            onClick: onConfirm,
            loading,
            children: confirmButtonText
          }
        )
      ] })
    ] });
  }
);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Popup
});
