// @wordpress/ui/build/tabs/context.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// packages/ui/src/tabs/context.tsx
var context_exports = {};
__export(context_exports, {
  TabsValidationProvider: () => TabsValidationProvider,
  useRegisterPanel: () => useRegisterPanel,
  useRegisterTab: () => useRegisterTab
});
module.exports = __toCommonJS(context_exports);
var import_element = require("@wordpress/element");
var import_jsx_runtime = require("react/jsx-runtime");
var VALIDATION_ENABLED = process.env.NODE_ENV !== "production";
var TabsValidationContext = VALIDATION_ENABLED ? (0, import_element.createContext)(null) : null;
function useRegisterTabDev() {
  const context = (0, import_element.useContext)(TabsValidationContext);
  (0, import_element.useEffect)(() => {
    if (context) {
      return context.registerTab();
    }
    return void 0;
  }, [context]);
}
function useRegisterTabProd() {
}
var useRegisterTab = VALIDATION_ENABLED ? useRegisterTabDev : useRegisterTabProd;
function useRegisterPanelDev() {
  const context = (0, import_element.useContext)(TabsValidationContext);
  (0, import_element.useEffect)(() => {
    if (context) {
      return context.registerPanel();
    }
    return void 0;
  }, [context]);
}
function useRegisterPanelProd() {
}
var useRegisterPanel = VALIDATION_ENABLED ? useRegisterPanelDev : useRegisterPanelProd;
function TabsValidationProviderDev({
  children
}) {
  const tabCountRef = (0, import_element.useRef)(0);
  const panelCountRef = (0, import_element.useRef)(0);
  const validationScheduledRef = (0, import_element.useRef)(null);
  const scheduleValidation = (0, import_element.useCallback)(() => {
    if (validationScheduledRef.current) {
      clearTimeout(validationScheduledRef.current);
    }
    validationScheduledRef.current = setTimeout(() => {
      const tabCount = tabCountRef.current;
      const panelCount = panelCountRef.current;
      if (tabCount !== panelCount) {
        throw new Error(
          `Tabs: Tab/Panel count mismatch (${tabCount} Tabs, ${panelCount} Panels). Each Tab must be associated with exactly one Panel. Mismatched or missing associations can break screen reader navigation and violate WAI-ARIA Tabs pattern requirements.`
        );
      }
      validationScheduledRef.current = null;
    }, 0);
  }, []);
  const registerTab = (0, import_element.useCallback)(() => {
    tabCountRef.current += 1;
    scheduleValidation();
    return () => {
      tabCountRef.current -= 1;
      scheduleValidation();
    };
  }, [scheduleValidation]);
  const registerPanel = (0, import_element.useCallback)(() => {
    panelCountRef.current += 1;
    scheduleValidation();
    return () => {
      panelCountRef.current -= 1;
      scheduleValidation();
    };
  }, [scheduleValidation]);
  (0, import_element.useEffect)(() => {
    return () => {
      if (validationScheduledRef.current) {
        clearTimeout(validationScheduledRef.current);
      }
    };
  }, []);
  const contextValue = (0, import_element.useMemo)(
    () => ({
      registerTab,
      registerPanel
    }),
    [registerTab, registerPanel]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TabsValidationContext.Provider, { value: contextValue, children });
}
function TabsValidationProviderProd({
  children
}) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_jsx_runtime.Fragment, { children });
}
var TabsValidationProvider = VALIDATION_ENABLED ? TabsValidationProviderDev : TabsValidationProviderProd;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  TabsValidationProvider,
  useRegisterPanel,
  useRegisterTab
});
